Documentation and tests for the forwarding routes [19249] (#398)
eProsima/DDS-Router: docs/rst/user_manual/configuration.rst

.. _user_manual_configuration_general_example:

Forwarding Routes
=================

The |ddsrouter| is capable of establishing different internal routes between its participants.

> This feature enables users to only send sensible data to a set of participants.

Reviewer: sensible?

> This feature enables users to only send sensible data to a set of participants.

Reviewer: ```suggestion
This feature enables users to only forward sensible data to a set of participants.
```

Reviewer: I would use "forward" instead everywhere, to avoid confusion with writers' send method.

.. note::

>     By default, when the tag ``routes`` is not set, every participant receives everything.

Reviewer: ```suggestion
    By default, when the tag ``routes`` is not set, every participant forwards received data to all others.
```

Generic Routes
--------------

> To configure a custom set of forwarding routes, use the tag ``routes`` followed by the destination participants of each source.

Reviewer: ```suggestion
To configure a custom set of forwarding routes, use the tag ``routes`` followed by the destination participants for each source participant.
```

.. note::

>     If a participant is not listed as a source, it will send the data it receives to every participant.

Reviewer: ```suggestion
    If a participant is not listed as a source, it will send the data it receives to every other participant.
```

.. note::

    If a participant is listed as a source but it is not given any destination participants, it will not send the data it receives to any participant.

Consider the following example with three participants: ``Participant0``, ``Participant1``, and ``Participant2``.

.. code-block:: yaml

  routes:
    - src: Participant0
      dst:
        - Participant2

    - src: Participant1

> When data is published:

Reviewer: ```suggestion
```

* Participant ``Participant0`` will only forward the data it receives to participant ``Participant2``.
* Participant ``Participant1`` will not forward the data it receives to any participant, since it does not have any destination participants.

> * Participant ``Participant2`` will forward the data it receives to every participant, since it does not have a forwarding route.

Reviewer: ```suggestion
* Participant ``Participant2`` will forward the data it receives to every participant (``Participant0`` and ``Participant1``), since it does not have a forwarding route.
```

.. warning::

>     A repeater participant with a route defined must add itself to its route's destinations.

Reviewer: Well said!

Author: Thank you

To configure a custom set of forwarding routes for a specific topic, use the tag ``topic-routes``.

.. warning::

  Topic routes take precedence over generic routes.
  This means that when data is published on a topic with a topic route configured, the generic routes are ignored and the topic route is used.

Again, consider the following example with three participants: ``Participant0``, ``Participant1``, and ``Participant2``.

.. code-block:: yaml

  topic-routes:
    - name: HelloWorld
      type: HelloWorld
      routes:
        - src: Participant1
          dst:
            - Participant0

        - src: Participant2

When data is published in the topic ``HelloWorld`` with type ``HelloWorld``:

* Participant ``Participant0`` will forward the data it receives to every participant, since it does not have a forwarding route.
* Participant ``Participant1`` will only forward the data it receives to participant ``Participant0``.
* Participant ``Participant2`` will not forward the data it receives to any participant, since it does not have any destination participants.

.. warning::

>   When configuring topic routes, the topic's ``type`` tag is required.

Reviewer: I would include this information at the introduction paragraph.

    ####################

>     # Simple DDS Participant in domain 3

Reviewer: ```suggestion
    # Simple DDS Participant in domain 7
```
